src/compiler/transformers/jsx.ts

```diff
@@ -24,7 +26,458 @@ namespace ts {
+
+                            addNode(properties, transformJsxAttributeToObjectLiteralElement(attr));
+                        }
+                    }
+
+                    if (properties) {
+                        segments.push(createObjectLiteral(properties));
+                    }
+
+                    objectProperties = createJsxSpread(compilerOptions.reactNamespace, segments);
+                }
+                else {
+                    const properties = map(node.attributes, transformJsxAttributeToObjectLiteralElement);
+                    objectProperties = createObjectLiteral(properties);
+                }
+            }
+
+            const childExpressions = filter(map(children, transformJsxChildToExpression), isDefined);
+            return createJsxCreateElement(compilerOptions.reactNamespace, tagName, objectProperties, childExpressions);
+        }
+
+        function transformJsxSpreadAttributeToExpression(node: JsxSpreadAttribute) {
+            return visitNode(node.expression, visitor, isExpression);
+        }
+
+        function transformJsxAttributeToObjectLiteralElement(node: JsxAttribute) {
+            const name = getAttributeName(node);
+            const expression = node.initializer
+                ? visitNode(node.initializer, visitor, isExpression)
+                : createLiteral(true);
+            return createPropertyAssignment(name, expression);
+        }
+
+        function visitJsxText(node: JsxText) {
+            const text = getTextToEmit(node);
+            if (text !== undefined) {
+                return createLiteral(text);
+            }
+            return undefined;
+        }
+
+        function getTextToEmit(node: JsxText) {
+            const text = trimReactWhitespaceAndApplyEntities(node);
+            if (text === undefined || text.length === 0) {
+                return undefined;
+            }
+            else {
+                return text;
+            }
+        }
+
+        function trimReactWhitespaceAndApplyEntities(node: JsxText): string {
+            const text = getTextOfNode(node, /*includeTrivia*/ true);
+            let result: string = undefined;
+            let firstNonWhitespace = 0;
+            let lastNonWhitespace = -1;
+
+            // JSX trims whitespace at the end and beginning of lines, except that the
+            // start/end of a tag is considered a start/end of a line only if that line is
+            // on the same line as the closing tag. See examples in tests/cases/conformance/jsx/tsxReactEmitWhitespace.tsx
+            for (let i = 0; i < text.length; i++) {
+                const c = text.charCodeAt(i);
+                if (isLineBreak(c)) {
+                    if (firstNonWhitespace !== -1 && (lastNonWhitespace - firstNonWhitespace + 1 > 0)) {
+                        const part = text.substr(firstNonWhitespace, lastNonWhitespace - firstNonWhitespace + 1);
+                        result = (result ? result + "\" + ' ' + \"" : "") + part;

```

Comment:
@rbuckton because it's much easier to make a source map for the former form, and it's what the canonical JSX transformer did at the time I wrote the emitter. Looks like Babel produces a concatenated string now; I don't have an opinion either way